### ISSUE TITLE
Unable to monitor

### ISSUE THREAD

**Issue by fonty422** [agerasev/pyepics-asyncio#1]
Running the default example:
```python
async with pv.monitor() as mon:
    async for value in mon:
        print(value)
```

yields an error:
```python
AttributeError: 'Pv' object has no attribute 'monitor'
```

Also, is there any possibility of adding connection callbacks?

**Comment by fonty422** [comment]
While at it, the basic `await pv.get()` appears to return None, while a classic pyepics `get()` on the same PV definitely give a value

**Comment by agerasev** [comment]
> Running the default example: `...` yields an error: `...`

Sorry, I forgot to update README to version 0.2. Updated now, thanks for pointing out.

In 0.2 there are two distinct Pv types: monitoring and non-monitoring. Here is a new monitoring example:
```python
from pyepics_asyncio import PvMonitor

pv = await PvMonitor.connect("pvname")
async for value in pv:
    print(value)
```

**Comment by agerasev** [comment]
> Also, is there any possibility of adding connection callbacks?

You can create basic pyepics `Pv` with desired parameters (e.g. connection callback) and then wrap it in pyepics_asyncio `Pv` or `PvMonitor`:

```python
def callback():
    ...

raw = epics.PV(
    "pvname",
    form="native",
    connection_callback=callback,
)

pv = Pv(raw)
```

**Comment by agerasev** [comment]
> While at it, the basic `await pv.get()` appears to return None, while a classic pyepics `get()` on the same PV definitely give a value

It's interesting. `Pv.get` simply [calls `epics.Pv.get` in executor and returns its result](https://github.com/agerasev/pyepics-asyncio/blob/master/pyepics_asyncio/__init__.py#L35-L39).
Could you show your code for both cases?

**Comment by fonty422** [comment]
Thanks heaps for the update. Here's the example you requested

```python
from pyepics_asyncio import Pv, PvMonitor
from epics import PV

pv = await Pv.connect("SR11BCM01:CURRENT_MONITOR")
pv_ = PV("SR11BCM01:CURRENT_MONITOR")

print(await pv.get())
print(pv_.get())

>> None
>> 200.19154430039114
```

I'll play around with the PvMonitor today and let you know if there are any further questions.

**Comment by fonty422** [comment]
It's still unhappy with things. I did also update to version 0.2.1 from 0.2.0 (which I assume was only the recent readme updates).
```python
pv_async = await PvMonitor.connect("SR11BCM01:CURRENT_MONITOR")
async for value in pv_async:
    print(value)
```

with the error being:
```python
----> 1 pv_async = await PvMonitor.connect("SR11BCM01:CURRENT_MONITOR")
AttributeError: '_ConnectMonitor' object has no attribute 'name'
```

**EDIT**
You can't monitor a PV that is referenced elsewhere? So if I have `pv = await Pv.connect("SR11BCM01:CURRENT_MONITOR")` or `pv_ = epics.PV("SR11BCM01:CURRENT_MONITOR")` I can't also have `pv_async = await PvMonitor.connect("SR11BCM01:CURRENT_MONITOR")`? If I only have the `pvMonitor` it works as expected.

The following still results in a None:
```python
from pyepics_asyncio import Pv, PvMonitor
pv = await Pv.connect("SR11BCM01:CURRENT_MONITOR")
print(await pv.get())
```

**Comment by fonty422** [comment]
Ah, the Pv class doesn't return the awaited value.
[This line](https://github.com/agerasev/pyepics-asyncio/blob/c941d59a057f5683bb68d5255f670ee90763f966/pyepics_asyncio/__init__.py#L39C9-L39C9) needs a return statement added

**Comment by agerasev** [comment]
Okay, I've write some [tests](https://github.com/agerasev/pyepics-asyncio/blob/v0.2/pyepics_asyncio/test_.py) with [dummy IOC](https://github.com/agerasev/pyepics-asyncio/blob/v0.2/ioc/TestApp/Db/Test.db) and found some issues:

1. `Pv.get` method wasn't returned value from `epics.PV.get()`. But I reworked it even further. PyEpics has no way to set callback for `PV.get`, so now `get` enables monitor and disables it after obtaining first value instead of using blocking `PV.get`.
2. Connection callback may fire before `_ConnectMonitor` future fully initialized. This leads to `object has no attribute 'name'` error. Fixed by calling `epics.PV.__init__` in-place only after future initialization.
3. It's not a bug, but undesirable behavior. `PvMonitor` previously stored all values it received, that caused memory overfill when being connected to frequently updated PV for a long time. Now monitor stores at most two values: last received and last unread.

Thank you for your assistance! It's nice to see that this library is used by someone.

**Comment by fonty422** [comment]
No, thank you for doing what I was about to try and do but wasn't sure I had the skill.
We have an asyncio application we're developing and I was just getting the event loop then using `call_soon_threadsafe()` on a basic callback to call another function, but your solution is much prettier.

I thinking I might try add a `_Notify` method to the PvMonitor so if that is not None, then it triggers the async call to perform some other action (to send to a kafka service, in our case). Would you prefer to keep that separate to this and keep the library clean and then extend this myself separately?
Oh, and we think the form option should be enabled as an option too, which can default to 'native' so we can get other parameters too.

**Comment by agerasev** [comment]
> I thinking I might try add a `_Notify` method to the PvMonitor so if that is not None, then it triggers the async call to perform some other action (to send to a kafka service, in our case). Would you prefer to keep that separate to this and keep the library clean and then extend this myself separately? Oh, and we think the form option should be enabled as an option too, which can default to 'native' so we can get other parameters too.

Feel free to open pull requests to this repo. If you need a feature, someone else may need it too.

As for notification callback, maybe it's better to simply spawn a new task that iterates over monitor? It seems to be more `async/await`-way rather than callbacks. Or I missing something?

**Comment by fonty422** [comment]
Ah, yes, you're right, the notify is more an async/await thing so should be separate.
I do have a few small ideas to make the package more flexible, but they're basically just allowing the user the option to tailor the request to include the full gamut of options from pyepics which can all default to some fall-back if they're not supplied.
I might pull and if you like the changes, then you're free to accept them.
